src/mcts/search.cc

```diff
@@ -1445,7 +1419,87 @@ void SearchWorker::DoBackupUpdateSingleNode(
   search_->total_playouts_ += node_to_process.multivisit;
   search_->cum_depth_ += node_to_process.depth * node_to_process.multivisit;
   search_->max_depth_ = std::max(search_->max_depth_, node_to_process.depth);
-}  // namespace lczero
+}
+
+void SearchWorker::MaybeConvertParent(Node* p, bool& can_convert,
+                                      float m) const {
+  // Convert parents to terminals except the root or those already converted.
+  can_convert = can_convert && p != search_->root_node_ && !p->IsTerminal();
+  if (!can_convert) return;
+
+  auto found_tb = false;
+  auto losing_m = 0.0f;
+
+  // Determine the maximum (Lower, Upper) bounds across all children.
+  // (-1,-1) Loss (initial and lowest bounds)
+  // (-1, 0) Can't Win
+  // (-1, 1) Regular node
+  // ( 0, 0) Draw
+  // ( 0, 1) Can't Lose
+  // ( 1, 1) Win (highest bounds)
+  auto max_L = -1;
+  auto max_U = -1;
+  for (const auto& edge : p->Edges()) {
+    if (edge.IsTerminal()) {
+      found_tb = found_tb || edge.IsTbTerminal();
+
+      const auto WL = edge.GetWL();
+      // We can short-circuit after a single win maxes the bounds.
+      if (WL > 0.0f) {
+        max_L = max_U = 1;
+        break;
+      } else if (WL == 0.0f) {
+        max_L = std::max(0, max_L);
+        max_U = std::max(0, max_U);
+      } else {
+        // Track the longest loss.
+        losing_m = std::max(losing_m, edge.GetM(0.0f));
+      }
+    } else if (edge.CantWin()) {
+      max_U = std::max(0, max_U);
+    } else if (edge.CantLose()) {
+      max_L = std::max(0, max_L);
+      max_U = 1;
```

Comment:
Why is it so? It doesn't mean this edge can win, does it?

Reply:
Which part are you asking about? Specifically that `max_U = 1` statement? `CantLose` bounds are `(0, 1)` so it'll max the U to 1 like a Win or Regular node.

`CantLose` is introduced when all possible moves are `Loss` or `CantWin`, e.g., your opponent can choose between a capture to tablebase loss or play into a position where you can force a draw. This means before that opponent position, you are in a situation where you `CantLose` because in the worst case, you can play into the draw -- although you might play a different move that does lose similar to how your opponent could suboptimally play into the tablebase loss.

So pretending there was a single edge that's `CantLose`, the parent would then have a lower bound of `-max_U == -1` and upper bound `-max_L == 0` i.e., parent `CantWin`.

Reply:
Oh, I think I see what you were asking? "If the edge CantLose then why can it win with `max_U = 1`" ? The max_L and max_U values are from the perspective of the children, so in this case, processing a CantLose means we can increase the lower bound up to 0 while increasing the upper bound to the max 1.

The comment block below shows how these bounds are inverted and reversed for the parent.